File: sys/helpers/readTextGrid.py
from __future__ import division
import sys, re

point_buffer = 0.025

rawfields = []
tgf = open('/Users/kate/speech-tech/voice-sauce/vs-octave/tests/sounds/textgrid/hmong_f4_40_a.TextGrid', 'rb')
for line in tgf:
	rawfields.append(line)
tgf.close()

tiers = 0
proceed_int = 0
proceed_pnt = 0
xmin = 0
xmax = 1

"""
cases: 'intervals:size, xmin, xmax, text, points:size, time, mark'
"""

fields = {
	'file_type': re.compile(r'(File\stype\s)=(?P<file_type>\s.+)'),
	'obj_class': re.compile(r'(Object\sclass\s)=(?P<obj_class>\s.+)'),
	'xmin': re.compile(r'(xmin\s)=(?P<xmin>\s\d+)'),
	'xmax': re.compile(r'(xmax\s)=(?P<xmax>\s\d+)'),
	'tiers?': re.compile(r'(tiers\?\s)(?P<tiers>\<.+)'),
	'item_1': re.compile(r'(\s+item\s)(?P<item_1>\[1.+)'),
	'item_class': re.compile(r'(\s+class\s=)(?P<item_class>\s+.+)'),
	'item_name': re.compile(r'(\s+name\s=)(?P<item_name>\s+.+)'),
	'intervals_size': re.compile(r'(\s+intervals\:\ssize\s=)(?P<intervals_size>\s\d)'),
	'interval_no': re.compile(r'(\s+intervals\s\[)(?P<interval_no>\d)\]'),
	'int_xmin': re.compile(r'(\s+xmin\s=)(?P<int_xmin>\s\d).*'),
	'int_xmax': re.compile(r'(\s+xmax\s=)(?P<int_xmax>\s\d).*'),
	'int_text': re.compile(r'(\s+text\s=)(?P<int_text>\w+)')
}

for f in rawfields:
	for k in fields.keys():
		m = re.match(fields[k], f)
		if m:
			print "match"
			print m.groupdict()

# test = rawfields[1]
# m = re.match(fields['obj_class'], test)
# if m:
# 	print m.group('obj_class')

# for f in rawfields:
# 	if any(re.match(fields[k], f) for k in fields)
# 	if m:
# 		print m.groups()







	

File: src/main/helpers/readTextGrid.m
function [labels, start, stop] = readTextGrid(filename)
	tgf = '';
	if(nargin == 0)
		tgf = '/Users/kate/speech-tech/voice-sauce/vs-octave/tests/sounds/textgrid/hmong_f4_40_a.TextGrid';
	else
		tgf = filename;
	end

	POINT_BUFFER = 0.025;

	assert (exist(tgf, 'file') ~= 0, 'error: text grid file [%s] not found', tgf);

	fid = fopen(tgf, 'rt');
	C = textscan(fid, '%s', 'delimiter', '\n');
	fclose(fid);

	C = C{1};
	tiers = 0;
	tier_len = 0;
	proceed_int = 0;
	proceed_pnt = 0;
	xmin = 0;
	xmax = 1;
	cnt = 0;

	for k=1:length(C)
		%printf('C{K} = [[  %s  ]]\n\n', C{k});

		if (isempty(C{k}))
			continue;
		end

		[m, t] = regexp(C{k}, '\s*(intervals[:]\ssize)\s=\s(\d)', 'match', 'tokens'); 
		if (~isempty(m)) % found start of new tier
			%printf('new interval:\n');
			t = t{1};
			tiers = tiers + 1;
			tier_len = str2num(t{2});
			labels{tiers} = cell(tier_len, 1);
			start{tiers} = zeros(tier_len, 1);
			stop{tiers} = zeros(tier_len, 1);
			cnt = 1;
			proceed_int = 1;
		end

		[m, t] = regexp(C{k}, '\s*(xmax|xmin)\s=\s([.\d]+)', 'match', 'tokens');
		if (~isempty(m)) % found xmax
			t = t{1};
			fname = t{1};
			val = t{2};
%			printf('fname = "%s"\n', fname);
%			printf('val = "%s"\n', val);
			xm = 0;
			if (strcmp(val, '0'))
				xm = 0.0;
				%printf('==> %s val == 0\n', fname);
			else
				xm = str2double(val);
			end

			switch fname
			case 'xmin'
				if (proceed_int)
					start{tiers}(cnt) = xm;
				else
					xmin = xm;
				end
			case 'xmax'
				if (proceed_int == 1)
					stop{tiers}(cnt) = xm;
				else
					xmax = xm;
				end
			end
		end

		[m, t] = regexp(C{k}, '\s*(text)\s=\s(["]\w*["])', 'match', 'tokens');
		if (~isempty(m))
			t = t{1};
			label = t{2};
			%printf('label = [%s]\n', label);
			labels{tiers}{cnt} = label;
			cnt = cnt + 1;
			if (cnt > tier_len)
				proceed_int = 0;
			end
		end

		[m, t] = regexp(C{k}, '\s*(points[:]\ssize)\s=\s(\d)', 'match', 'tokens');
		if (~isempty(m))
			printf('USING POINT-SOURCES: NEED TO DEBUG THIS.\n\n');
			t = t{1};
			tiers = tiers + 1;
			tier_len = str2num(t{2});
			labels{tiers} = cell(tier_len, 1);
			start{tiers} = zeros(tier_len, 1);
			stop{tiers} = zeros(tier_len, 1);
			cnt = 1;
			proceed_pnt = 1;
		end

		[m, t] = regexp(C{k}, '.*(time).*=.*([.\d]+)', 'match', 'tokens');
		if (~isempty(m))
			t = t{1};
			val = num2double(t{2});
			if (proceed_pnt == 1)
				start{tiers}(cnt) = val - POINT_BUFFER;
				stop{tiers}(cnt) = val + POINT_BUFFER;

				if (start{tiers}(cnt) < xmin)
					start{tiers}(cnt) = xmin;
				end

				if (stop{tiers}(cnt) > xmax)
					stop{tiers}(cnt) = xmax;
				end
			end
		end

		[m, t] = regexp(C{k}, '.*(mark).*=.*(["]?\w+["]?)', 'match', 'tokens');
		if (~isempty(m))
			t = t{1};
			label = t{2};
			labels{tiers}{cnt} = label;
			cnt = cnt + 1;
			if (cnt > tier_len)
				proceed_pnt = 0;
			end
		end
	end
end



	
	

